### PR TITLE
[func](jsonb) support json_contains_path function

### PR DESCRIPTION
**Pull request by zy-kkk**
## Proposed changes

Issue Number: close #xxx

#24366

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zy-kkk** [comment]
run buildall

**Comment by xiaokang** [comment]
run buildall

File: be/src/vec/functions/function_jsonb.cpp [hunk 1]
```diff
@@ -1288,6 +1288,188 @@ struct JsonbContainsAndPathImpl {
...
+                               size_t input_rows_count) {
+        auto path = ColumnString::create();
+        std::string root_path = "$";
+
+        for (int i = 0; i < input_rows_count; i++) {
+            reinterpret_cast<ColumnString*>(path.get())
+                    ->insert_data(root_path.data(), root_path.size());
+        }
+
+        block.insert({std::move(path), std::make_shared<DataTypeString>(), "path"});
```
Comment (xiaokang) [inline be/src/vec/functions/function_jsonb.cpp:1450]: Where is the path column used?

File: be/src/vec/functions/function_jsonb.cpp [hunk 2]
```diff
@@ -1353,6 +1535,8 @@ void register_function_jsonb(SimpleFunctionFactory& factory) {
     factory.register_function<FunctionJsonbLength<JsonbLengthAndPathImpl>>();
     factory.register_function<FunctionJsonbContains<JsonbContainsImpl>>();
     factory.register_function<FunctionJsonbContains<JsonbContainsAndPathImpl>>();
+    factory.register_function<FunctionJsonbContainsPath<JsonbContainsPathWithOnePathImpl>>();
+    factory.register_function<FunctionJsonbContainsPath<JsonbContainsPathWithTwoPathImpl>>();
```
[the remark this answers is not in the document]
Comment (xiaokang) [reply be/src/vec/functions/function_jsonb.cpp:1539]: `[JSON_CONTAINS_PATH(json_doc, one_or_all, path[, path] ...)](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-contains-path)`

mysql ref: https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-contains-path

File: be/src/vec/functions/function_jsonb.cpp (hunk 1 above)
Comment (xiaokang) [reply be/src/vec/functions/function_jsonb.cpp:1450]: What's the purpose of path '$'?

**Comment by zy-kkk** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/JsonContainsPath.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.doris.nereids.types.VarcharType;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * ScalarFunction 'json_contains_path'. This class is generated by GenerateFunction .
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/JsonContainsPath.java:40]: update this comment to how to use it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/JsonContainsPath.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    @Override
+    public List<FunctionSignature> getSignatures() {
+        String oneOrAll = ((StringLikeLiteral) child(1)).getStringValue().toLowerCase();
+        if (!oneOrAll.equals("one") && !oneOrAll.equals("all")) {
+            throw new AnalysisException(
+                    "The oneOrAll argument to json_contains_path may take these values: 'one' or 'all'.");
+        } else {
+            return SIGNATURES;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/JsonContainsPath.java:76]: if u assume the second arg is `StringLikeLiteral`. u should impl `checkLegalityBeforeTypeCoercion` and check the arg in it

File: be/src/vec/functions/function_jsonb.cpp [hunk 5]
```diff
@@ -1288,6 +1288,188 @@ struct JsonbContainsAndPathImpl {
...
+                    return Status::InvalidArgument(
+                            "Json path error: {} for value: {}",
+                            JsonbErrMsg::getErrMsg(JsonbErrType::E_INVALID_JSON_PATH),
+                            std::string_view(reinterpret_cast<const char*>(path2_value.data),
+                                             path2_value.size));
+                }
+            }
+
+            auto jsonb_value = jsonb_data_column->get_data_at(i);
+            auto one_or_all_value = to_lower(one_or_all_column->get_data_at(i).to_string());
```
[the remark this answers is not in the document]
Comment (xiaokang) [reply be/src/vec/functions/function_jsonb.cpp:1391]: If one_or_all_column is const, just check its value 'one' or 'all' once

File: be/src/vec/functions/function_jsonb.cpp [hunk 6]
```diff
@@ -1287,6 +1289,157 @@ struct JsonbContainsAndPathImpl {
...
+    bool use_default_implementation_for_constants() const override { return false; }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) const override {
+        return Impl::execute_impl(context, block, arguments, result, input_rows_count);
+    }
+};
+
+struct JsonbContainsPathUtil {
+    static Status jsonb_contains_path_execute(FunctionContext* context, Block& block,
```
Comment (xiaokang) [inline be/src/vec/functions/function_jsonb.cpp:1327]: JsonbContainsPathUtil::jsonb_contains_path_execute is only used by JsonbContainsPathImpl, why add a new class JsonbContainsPathUtil?

File: be/src/vec/functions/function_jsonb.cpp [hunk 7]
```diff
@@ -1287,6 +1289,157 @@ struct JsonbContainsAndPathImpl {
...
+                }
+            }
+        }
+
+        const auto return_type = block.get_data_type(result);
+        MutableColumnPtr res = return_type->create_column();
+
+        for (size_t i = 0; i < input_rows_count; ++i) {
+            if (jsonb_data_column->is_null_at(i) || one_or_all_column->is_null_at(i) ||
+                any_path_null) {
```
Comment (xiaokang) [inline be/src/vec/functions/function_jsonb.cpp:1372]: for type 'one', if some path is null, should the result be null?

**Pull request closed without merging**